# INTRODUCTION

Selon l'Organisation Mondiale de la Santé18 (OMS), le cancer du sein est le cancer le plus fréquent chez les femmes dans les pays développés et les moins développés. En 2018, près de 2 millions de nouveaux cas de cancer du sein ont été diagnostiqués. Les médecins savent que le cancer du sein survient lorsque certaines cellules mammaires commencent à se développer anormalement. Ces cellules se divisent plus rapidement que les cellules saines et continuent de s'accumuler, formant une masse. Le problème avec ces cellules cancérigènes est qu'elles ne causent pas de douleur ou d'inconfort jusqu'à ce qu'elles se soient propagées aux tissus voisins.
Dès lors, la détection d'un cancer du sein à un stade peu avancé de son développement peut permettre de soigner plus facilement mais aussi de limiter les séquelles liées à certains traitements.
Dans cette optique d'amélioration des diagnostics et des soins préventifs, l'Intelligence Artificielle peut aider les professionnels de Santé dans leur prise de décisions et permettre aux patients d'être traités rapidement. Les hôpitaux collectent de plus en plus de données dont l'analyse est un défi et une opportunité, Plusieurs Organisations (Centres de recherche, Start-ups, Entreprises, etc.) à travers le monde travaillent sur le développement d'algorithmes d'apprentissage automatique qui sont entraînés par les données collectées. Les modèles ainsi créés sont ensuite utilisés pour prédire par exemple quand une insuffisance respiratoire, une septicémie ou d'autres problèmes tels que le cancer du sein sont susceptibles de se produire.

Dans ce projet, nous verrons un cas d'utilisation du Machine Learning dans la Santé. Notre objectif est de construire un modèle capable de prédire si une patiente, sur la base de ses données, sera atteinte d'un cancer du sein ou pas.



# LIBRAIRIES

In [1]:
import pandas as pd
import numpy as np

from sklearn.datasets import load_breast_cancer
#from sklearn.neighbors import KNeighborsClassifier
from sklearn.model_selection import cross_val_score, GridSearchCV

DONNEES

Les données utilisées dans le cadre de ce projet, proviennent de UCI Machine Learning Repository. 1º Il s'agit de la base de données diagnostic du cancer du sein du Wisconsin aux USA. Ces données sont également disponibles dans le module datasets.load breast cancer de sklearn.20 Veuillez consulter la page descriptive " de ces données pour une meilleure compréhension.

In [3]:
# Importantion des données 
df = load_breast_cancer()

In [5]:
# Affichage ders données
print(df)

{'data': array([[1.799e+01, 1.038e+01, 1.228e+02, ..., 2.654e-01, 4.601e-01,
        1.189e-01],
       [2.057e+01, 1.777e+01, 1.329e+02, ..., 1.860e-01, 2.750e-01,
        8.902e-02],
       [1.969e+01, 2.125e+01, 1.300e+02, ..., 2.430e-01, 3.613e-01,
        8.758e-02],
       ...,
       [1.660e+01, 2.808e+01, 1.083e+02, ..., 1.418e-01, 2.218e-01,
        7.820e-02],
       [2.060e+01, 2.933e+01, 1.401e+02, ..., 2.650e-01, 4.087e-01,
        1.240e-01],
       [7.760e+00, 2.454e+01, 4.792e+01, ..., 0.000e+00, 2.871e-01,
        7.039e-02]]), 'target': array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0,
       0, 0, 1, 0, 1, 1, 1, 1, 1, 0, 0, 1, 0, 0, 1, 1, 1, 1, 0, 1, 0, 0,
       1, 1, 1, 1, 0, 1, 0, 0, 1, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 0, 0,
       1, 1, 1, 0, 1, 1, 0, 0, 1, 1, 1, 0, 0, 1, 1, 1, 1, 0, 1, 1, 0, 1,
       1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 1, 0,
 





df est un dictionnaire contenant plusieurs paires clés-valeurs. La clé « data » a pour valeurs un tableau numpy des données des variables indépendantes. La clé « target » a pour valeur un tableau numpy représentant la variable cible. Cette variable cible a deux modalités malignant*
(1) et benign' (O). La clé « features names » contient un tableau numpy des noms des variables indépendantes.


In [8]:
# Variable cible
y = df.target

In [10]:
print(y)

[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 1 0 0 0 0 0 0 0 0 1 0 1 1 1 1 1 0 0 1 0 0 1 1 1 1 0 1 0 0 1 1 1 1 0 1 0 0
 1 0 1 0 0 1 1 1 0 0 1 0 0 0 1 1 1 0 1 1 0 0 1 1 1 0 0 1 1 1 1 0 1 1 0 1 1
 1 1 1 1 1 1 0 0 0 1 0 0 1 1 1 0 0 1 0 1 0 0 1 0 0 1 1 0 1 1 0 1 1 1 1 0 1
 1 1 1 1 1 1 1 1 0 1 1 1 1 0 0 1 0 1 1 0 0 1 1 0 0 1 1 1 1 0 1 1 0 0 0 1 0
 1 0 1 1 1 0 1 1 0 0 1 0 0 0 0 1 0 0 0 1 0 1 0 1 1 0 1 0 0 0 0 1 1 0 0 1 1
 1 0 1 1 1 1 1 0 0 1 1 0 1 1 0 0 1 0 1 1 1 1 0 1 1 1 1 1 0 1 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 1 1 1 1 1 1 0 1 0 1 1 0 1 1 0 1 0 0 1 1 1 1 1 1 1 1 1 1 1 1
 1 0 1 1 0 1 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 0 1 1 1 0 1 0 1 1 1 1 0 0 0 1 1
 1 1 0 1 0 1 0 1 1 1 0 1 1 1 1 1 1 1 0 0 0 1 1 1 1 1 1 1 1 1 1 1 0 0 1 0 0
 0 1 0 0 1 1 1 1 1 0 1 1 1 1 1 0 1 1 1 0 1 1 0 0 1 1 1 1 1 1 0 1 1 1 1 1 1
 1 0 1 1 1 1 1 0 1 1 0 1 1 1 1 1 1 1 1 1 1 1 1 0 1 0 0 1 0 1 1 1 1 1 0 1 1
 0 1 0 1 1 0 1 0 1 1 1 1 1 1 1 1 0 0 1 1 1 1 1 1 0 1 1 1 1 1 1 1 1 1 1 0 1
 1 1 1 1 1 1 0 1 0 1 1 0 

In [17]:
y.shape

(569,)

In [11]:
# Variables indépendantes

X = df.data

In [15]:
X

array([[1.799e+01, 1.038e+01, 1.228e+02, ..., 2.654e-01, 4.601e-01,
        1.189e-01],
       [2.057e+01, 1.777e+01, 1.329e+02, ..., 1.860e-01, 2.750e-01,
        8.902e-02],
       [1.969e+01, 2.125e+01, 1.300e+02, ..., 2.430e-01, 3.613e-01,
        8.758e-02],
       ...,
       [1.660e+01, 2.808e+01, 1.083e+02, ..., 1.418e-01, 2.218e-01,
        7.820e-02],
       [2.060e+01, 2.933e+01, 1.401e+02, ..., 2.650e-01, 4.087e-01,
        1.240e-01],
       [7.760e+00, 2.454e+01, 4.792e+01, ..., 0.000e+00, 2.871e-01,
        7.039e-02]])

In [18]:
print(X)

[[1.799e+01 1.038e+01 1.228e+02 ... 2.654e-01 4.601e-01 1.189e-01]
 [2.057e+01 1.777e+01 1.329e+02 ... 1.860e-01 2.750e-01 8.902e-02]
 [1.969e+01 2.125e+01 1.300e+02 ... 2.430e-01 3.613e-01 8.758e-02]
 ...
 [1.660e+01 2.808e+01 1.083e+02 ... 1.418e-01 2.218e-01 7.820e-02]
 [2.060e+01 2.933e+01 1.401e+02 ... 2.650e-01 4.087e-01 1.240e-01]
 [7.760e+00 2.454e+01 4.792e+01 ... 0.000e+00 2.871e-01 7.039e-02]]


In [20]:
X.shape

(569, 30)

# MODELISATION

Nous sommes face à un problème de classification. 

Parmi les algorithmes de classification les plus utilisés en Data Science, il y a l'algorithme des k plus proches voisins (K Nearest Neighbor ou KNN en Anglais). KNN est un algorithme d'apprentissage automatique supervisé particulièrement simple et facile à implémenter.

Commençons d'abord par l'initialisation de l'algorithme KNN avec le module
sklearn.neighbors.KNeighborsClassifier et ses paramètres par défaut :

In [21]:
from sklearn.neighbors import KNeighborsClassifier

In [24]:
# Initialisation de KNN avec ses hyperparamètres par défaut
#knn = KNeighborsClassifier ()

knn = KNeighborsClassifier()

In [25]:
knn

KNeighborsClassifier()

Réalisons ensuite une validation croisée avec 5 plis :

In [29]:
# 5-fold cross-validation

cv_Scores = cross_val_score(knn,
                           X,
                           y,
                           cv=5,
                           scoring = 'precision')

# il ya 5 plis

In [30]:
# Affichage des scores

print(cv_Scores)

[0.8625     0.93243243 0.94520548 0.94594595 0.95652174]


Interprétation des Valeurs Individuelles
Chaque valeur représente la précision du modèle sur un pli particulier utilisé comme jeu de test pendant la validation croisée.

0.8625 : Précision du modèle sur le premier pli de test.

0.93243243 : Précision du modèle sur le deuxième pli de test.

0.94520548 : Précision du modèle sur le troisième pli de test.

0.94594595 : Précision du modèle sur le quatrième pli de test.

0.95652174 : Précision du modèle sur le cinquième pli de test.

Précision
La précision est une mesure de performance pour les modèles de classification. Elle est définie comme le ratio des vrais positifs (TP) sur le nombre total de prédictions positives (TP + FP) :


Precision= Vrais Positifs (TP)/Vrais Positifs (TP)+Faux Positifs (FP)


 

Une précision de 0.8625 signifie que 86.25% des prédictions positives faites par le modèle étaient correctes pour le premier pli de test. De même, pour le deuxième pli de test, 93.24% des prédictions positives étaient correctes, et ainsi de suite.

Performance du Modèle

Calculons le score moyen :

In [31]:
# Score moyen arrondi à deux chiffres après la virgule

round(np.mean(cv_Scores), 2)

0.93

Si je comprends bien. Lorsqu'on a une chaîne de charactère on met de guillemet.
et Au contraire si nous avons une variable numérique on ne met pas de guillemet
oui exactement!

Le score moyen est égal à 0,93.

# GRID SEARCH

Pour effectuer notre validation croisée, nous avons considéré les paramètres par défaut de l'algorithme de KNN : ce sont les hyperparamètres. Les hyperparamètres sont définis avant le début du processus de modélisation. Il ne faut pas les confondre avec les paramètres d'un modèle obtenus après l'entraînement de l'algorithme par des données. Pour l'algorithme de classification KNN dans Scikit-Learn, les hyperparamètres sont :

In [32]:
# Affichage de Knn

print(knn)

KNeighborsClassifier()


In [33]:
KNeighborsClassifier(algorithm='auto',
                    leaf_size=30,
                    metric_params= None,
                    n_jobs= None,
                    n_neighbors = 5,
                    weights= 'uniform')

KNeighborsClassifier()

Veuillez consulter cette documentation pour les définitions de ces algorithmes. Par exemple n neighbors indique le nombre de points voisins que l'algorithme considère dans ses calculs.
Comme vous le voyez bien, des valeurs par défaut sont définies pour chaque hyperparamètre.

Comment pouvons-nous être sûr que ces valeurs par défaut sont optimales pour notre modèle ?
Quelle est la meilleure combinaison des hyperparamètres qui donne la meilleure performance du modèle ? Imaginons que nous voulions essayer plusieurs valeurs pour les hyperparamètres algorithm, leaf _size, et n neighbors. Il faudra donc essayer chaque combinaison de ces valeurs, construire le modèle et l'évaluer par une validation croisée pour finalement choisir l'hyperparamétrage qui donne la meilleure métrique d'évaluation. Ce processus peut être effectué automatiquement avec une recherche de grille (Grid Search). Voici comment nous l'implémentons dans Python :


In [34]:
# Dictionnaire des hyperparamètres avec leurs valeurs

grid = {'n_neighbors': [1, 3, 5, 7, 9, 10 ],
       'weights' : ["uniform", "distance"],
       "algorithm": ['auto', 'ball_tree', 'kd_tree', 'brute'],
       "leaf_size": [10, 20, 30,40, 50]}

In [35]:
from sklearn.model_selection import GridSearchCV

In [38]:
# Estimateur
model = GridSearchCV(estimator= KNeighborsClassifier(),
                    param_grid= grid,
                    scoring= 'precision',
                    cv = 5)

In [41]:
model

GridSearchCV(cv=5, estimator=KNeighborsClassifier(),
             param_grid={'algorithm': ['auto', 'ball_tree', 'kd_tree', 'brute'],
                         'leaf_size': [10, 20, 30, 40, 50],
                         'n_neighbors': [1, 3, 5, 7, 9, 10],
                         'weights': ['uniform', 'distance']},
             scoring='precision')

Nous avons d'abord défini un dictionnaire avec comme clés les hyperparamètres et leurs valeurs que nous voulons essayer (veiller à bien écrire le nom de chaque hyperparamètre). Nous avons  ensuite défini un objet GridSearchCV qui nous permettra de rechercher le meilleur hyperparamétrage. Afin de mieux comprendre, calculons le nombre total de modèles qui seront construit. Chaque combinaison de valeur sera essayée. Pour une combinaison, 5 modèles seront construits (cv =5) afin de choisir celui qui donne la meilleure métrique (ici 'precision'). 

Pour chaque valeur d'un hyperparamètre, nous testons chaque valeur de chaque autre hyperparamètre : c'est ce qui forme une combinaison. Pour trouver le nombre de combinaisons possibles, il faudra donc multiplier les longueurs des listes qui définissent les valeurs des hyperparamètres. 

Alors, dans notre cas, le nombre de combinaisons possibles est égal à 6 x 2 x 4 x 2 x 5 soit 480 combinaisons. 5 modèles sont créés pour chaque combinaison. Alors le nombre total de modèles créés est égal à 480 x 5 soit 2400 modèles.

In [42]:
# Recherche du meilleur modèle
model.fit(X, y)

GridSearchCV(cv=5, estimator=KNeighborsClassifier(),
             param_grid={'algorithm': ['auto', 'ball_tree', 'kd_tree', 'brute'],
                         'leaf_size': [10, 20, 30, 40, 50],
                         'n_neighbors': [1, 3, 5, 7, 9, 10],
                         'weights': ['uniform', 'distance']},
             scoring='precision')

L'objet GridSearchCV a un attribut « cv_ results_» qui stocke dans un dictionnaire les résultats de la recherche. Nous allons affiché ces résultats dans un format plus lisible qu'un dictionnaire :
une dataframe.

In [44]:
# Convertion du dictionnaire des résultats en une dataframe
results = pd.DataFrame(model.cv_results_)

In [45]:
results

mean_fit_time  std_fit_time  mean_score_time  std_score_time  \
0              0.0           0.0         0.015705        0.000164   
1              0.0           0.0         0.021957        0.007753   
2              0.0           0.0         0.022035        0.007530   
3              0.0           0.0         0.021954        0.007748   
4              0.0           0.0         0.021959        0.005944   
..             ...           ...              ...             ...   
235            0.0           0.0         0.021954        0.007748   
236            0.0           0.0         0.018829        0.006211   
237            0.0           0.0         0.018836        0.005042   
238            0.0           0.0         0.018835        0.008351   
239            0.0           0.0         0.015705        0.000157   

    param_algorithm param_leaf_size param_n_neighbors param_weights  \
0              auto              10                 1       uniform   
1              auto              10                 1      distance   
2              auto              10                 3       uniform   
3              auto              10                 3      distance   
4              auto              10                 5       uniform   
..              ...             ...               ...           ...   
235           brute              50                 7      distance   
236           brute              50                 9       uniform   
237           brute              50                 9      distance   
238           brute              50                10       uniform   
239           brute              50                10      distance   

                                                params  split0_test_score  \
0    {'algorithm': 'auto', 'leaf_size': 10, 'n_neig...           0.857143   
1    {'algorithm': 'auto', 'leaf_size': 10, 'n_neig...           0.857143   
2    {'algorithm': 'auto', 'leaf_size': 10, 'n_neig...           0.851852   
3    {'algorithm': 'auto', 'leaf_size': 10, 'n_neig...           0.851852   
4    {'algorithm': 'auto', 'leaf_size': 10, 'n_neig...           0.862500   
..                                                 ...                ...   
235  {'algorithm': 'brute', 'leaf_size': 50, 'n_nei...           0.841463   
236  {'algorithm': 'brute', 'leaf_size': 50, 'n_nei...           0.851852   
237  {'algorithm': 'brute', 'leaf_size': 50, 'n_nei...           0.841463   
238  {'algorithm': 'brute', 'leaf_size': 50, 'n_nei...           0.851852   
239  {'algorithm': 'brute', 'leaf_size': 50, 'n_nei...           0.841463   

     split1_test_score  split2_test_score  split3_test_score  \
0             0.920000           0.918919           0.930556   
1             0.920000           0.918919           0.930556   
2             0.930556           0.945946           0.945205   
3             0.931507           0.945946           0.958333   
4             0.932432           0.945205           0.945946   
..                 ...                ...                ...   
235           0.932432           0.945946           0.945946   
236           0.932432           0.945946           0.946667   
237           0.932432           0.945946           0.958904   
238           0.932432           0.945946           0.958904   
239           0.932432           0.945946           0.958904   

     split4_test_score  mean_test_score  std_test_score  rank_test_score  
0             0.929577         0.911239        0.027465              201  
1             0.929577         0.911239        0.027465              201  
2             0.942029         0.923118        0.036058              161  
3             0.942857         0.926099        0.038094              101  
4             0.956522         0.928521        0.033882               61  
..                 ...              ...             ...              ...  
235           0.930556         0.919269        0.039440              181  
236           0.970588       

In [46]:
# 'params' et 'mean_test score' sont les attributs qui nous interessent
best = results[["params", 'mean_test_score']].sort_values(by = "mean_test_score", ascending= False).head()

In [47]:
best

params  mean_test_score
226  {'algorithm': 'brute', 'leaf_size': 40, 'n_nei...         0.931945
238  {'algorithm': 'brute', 'leaf_size': 50, 'n_nei...         0.931945
118  {'algorithm': 'ball_tree', 'leaf_size': 50, 'n...         0.931945
130  {'algorithm': 'kd_tree', 'leaf_size': 10, 'n_n...         0.931945
214  {'algorithm': 'brute', 'leaf_size': 30, 'n_nei...         0.931945

Les 5 meilleurs modèles ont le même score avec des hyperparamètres différents.

In [48]:
# Meilleur hyperparamétrage

best['params'].iloc[0]

{'algorithm': 'brute',
 'leaf_size': 40,
 'n_neighbors': 10,
 'weights': 'uniform'}

In [49]:
# Meilleur score
best_score = best['mean_test_score'].iloc[0]
print(best_score)

0.9319445150226777


Nous avons déterminé les valeurs des hyperparamètres ayant donné le meilleur score.
Malheureusement, ils n'ont pas amélioré le score trouvé avec les hyperparamètres par défaut qui était aussi égal à 0,93. Nous pouvons utiliser notre objet GridSearchCV (« model ») directement comme modèle de classification et l'utiliser pour effectuer des prédictions. Nous n'avons donc pas besoin d'entraîner à nouveau un algorithme de KNN avec les meilleurs hyperparamètres trouvés.

# CONCLUSION

Dans ce projet, nous avons construit un modèle de de prédiction du cancer de sein. De plus, nous avons appris comment rechercher les hyperparamètres optimaux du modèle. La méthode de Grid Search CV telle que décrite ici, peut aussi être appliquée pour n'importe quel projet de machine Learning aussi bien pour des tâches de classification que pour des tâches de régression et même en apprentissage non supervisé. Si cette méthode a l'avantage d'être un processus automatisé, facile à comprendre et à implémenter, il n'en demeure pas moins qu'elle a quelques inconvénients. L'inconvénient majeur de la méthode Grid Search est qu'elle est très coûteuse en calcul, c'est-à-dire que lorsque le nombre d'hyperparamètres à essayer augmente considérablement et que la dataframe est volumineuse, les temps de traitement peuvent être très lents. De plus, lorsque vous définissez votre grille, vous pouvez omettre par inadvertance un hyperparamétrage qui serait en fait optimal. S'il n'est pas spécifié dans votre grille, il ne sera jamais essayé !

Pour surmonter ces inconvénients, nous examinerons la recherche aléatoire (Random Search) dans le projet suivant.